// @dinero.js/currencies/dist/umd/latest.development.js
/*! @dinero.js/currencies 2.0.0-alpha.14 (UNRELEASED 1320a09) | MIT License | © Sarah Dayan and contributors | https://v2.dinerojs.com */
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
  typeof define === 'function' && define.amd ? define(['exports'], factory) :
  (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global["@dinero.js/currencies"] = {}));
})(this, (function (exports) { 'use strict';

  /**
   * United Arab Emirates dirham.
   */
  var AED = {
    code: 'AED',
    base: 10,
    exponent: 2
  };

  /**
   * Afghan afghani.
   */
  var AFN = {
    code: 'AFN',
    base: 10,
    exponent: 2
  };

  /**
   * Albanian lek.
   */
  var ALL = {
    code: 'ALL',
    base: 10,
    exponent: 2
  };

  /**
   * Armenian dram.
   */
  var AMD = {
    code: 'AMD',
    base: 10,
    exponent: 2
  };

  /**
   * Netherlands Antillean guilder.
   */
  var ANG = {
    code: 'ANG',
    base: 10,
    exponent: 2
  };

  /**
   * Angolan kwanza.
   */
  var AOA = {
    code: 'AOA',
    base: 10,
    exponent: 2
  };

  /**
   * Argentine peso.
   */
  var ARS = {
    code: 'ARS',
    base: 10,
    exponent: 2
  };

  /**
   * Australian dollar.
   */
  var AUD = {
    code: 'AUD',
    base: 10,
    exponent: 2
  };

  /**
   * Aruban florin.
   */
  var AWG = {
    code: 'AWG',
    base: 10,
    exponent: 2
  };

  /**
   * Azerbaijani manat.
   */
  var AZN = {
    code: 'AZN',
    base: 10,
    exponent: 2
  };

  /**
   * Bosnia and Herzegovina convertible mark.
   */
  var BAM = {
    code: 'BAM',
    base: 10,
    exponent: 2
  };

  /**
   * Barbados dollar.
   */
  var BBD = {
    code: 'BBD',
    base: 10,
    exponent: 2
  };

  /**
   * Bangladeshi taka.
   */
  var BDT = {
    code: 'BDT',
    base: 10,
    exponent: 2
  };

  /**
   * Bulgarian lev.
   */
  var BGN = {
    code: 'BGN',
    base: 10,
    exponent: 2
  };

  /**
   * Bahraini dinar.
   */
  var BHD = {
    code: 'BHD',
    base: 10,
    exponent: 3
  };

  /**
   * Burundian franc.
   */
  var BIF = {
    code: 'BIF',
    base: 10,
    exponent: 0
  };

  /**
   * Bermudian dollar.
   */
  var BMD = {
    code: 'BMD',
    base: 10,
    exponent: 2
  };

  /**
   * Brunei dollar.
   */
  var BND = {
    code: 'BND',
    base: 10,
    exponent: 2
  };

  /**
   * Bolivian boliviano.
   */
  var BOB = {
    code: 'BOB',
    base: 10,
    exponent: 2
  };

  /**
   * Bolivian Mvdol.
   */
  var BOV = {
    code: 'BOV',
    base: 10,
    exponent: 2
  };

  /**
   * Brazilian real.
   */
  var BRL = {
    code: 'BRL',
    base: 10,
    exponent: 2
  };

  /**
   * Bahamian dollar.
   */
  var BSD = {
    code: 'BSD',
    base: 10,
    exponent: 2
  };

  /**
   * Bhutanese ngultrum.
   */
  var BTN = {
    code: 'BTN',
    base: 10,
    exponent: 2
  };

  /**
   * Botswana pula.
   */
  var BWP = {
    code: 'BWP',
    base: 10,
    exponent: 2
  };

  /**
   * Belarusian ruble.
   */
  var BYN = {
    code: 'BYN',
    base: 10,
    exponent: 2
  };

  /**
   * Belize dollar.
   */
  var BZD = {
    code: 'BZD',
    base: 10,
    exponent: 2
  };

  /**
   * Canadian dollar.
   */
  var CAD = {
    code: 'CAD',
    base: 10,
    exponent: 2
  };

  /**
   * Congolese franc.
   */
  var CDF = {
    code: 'CDF',
    base: 10,
    exponent: 2
  };

  /**
   * WIR Euro.
   */
  var CHE = {
    code: 'CHE',
    base: 10,
    exponent: 2
  };

  /**
   * Swiss franc.
   */
  var CHF = {
    code: 'CHF',
    base: 10,
    exponent: 2
  };

  /**
   * WIR Franc.
   */
  var CHW = {
    code: 'CHW',
    base: 10,
    exponent: 2
  };

  /**
   * Unidad de Fomento.
   */
  var CLF = {
    code: 'CLF',
    base: 10,
    exponent: 4
  };

  /**
   * Chilean peso.
   */
  var CLP = {
    code: 'CLP',
    base: 10,
    exponent: 0
  };

  /**
   * Renminbi (Chinese) yuan.
   */
  var CNY = {
    code: 'CNY',
    base: 10,
    exponent: 2
  };

  /**
   * Colombian peso.
   */
  var COP = {
    code: 'COP',
    base: 10,
    exponent: 2
  };

  /**
   * Unidad de Valor Real.
   */
  var COU = {
    code: 'COU',
    base: 10,
    exponent: 2
  };

  /**
   * Costa Rican colón.
   */
  var CRC = {
    code: 'CRC',
    base: 10,
    exponent: 2
  };

  /**
   * Cuban convertible peso.
   */
  var CUC = {
    code: 'CUC',
    base: 10,
    exponent: 2
  };

  /**
   * Cuban peso.
   */
  var CUP = {
    code: 'CUP',
    base: 10,
    exponent: 2
  };

  /**
   * Cape Verdean escudo.
   */
  var CVE = {
    code: 'CVE',
    base: 10,
    exponent: 2
  };

  /**
   * Czech koruna.
   */
  var CZK = {
    code: 'CZK',
    base: 10,
    exponent: 2
  };

  /**
   * Djiboutian franc.
   */
  var DJF = {
    code: 'DJF',
    base: 10,
    exponent: 0
  };

  /**
   * Danish krone.
   */
  var DKK = {
    code: 'DKK',
    base: 10,
    exponent: 2
  };

  /**
   * Dominican peso.
   */
  var DOP = {
    code: 'DOP',
    base: 10,
    exponent: 2
  };

  /**
   * Algerian dinar.
   */
  var DZD = {
    code: 'DZD',
    base: 10,
    exponent: 2
  };

  /**
   * Egyptian pound.
   */
  var EGP = {
    code: 'EGP',
    base: 10,
    exponent: 2
  };

  /**
   * Eritrean nakfa.
   */
  var ERN = {
    code: 'ERN',
    base: 10,
    exponent: 2
  };

  /**
   * Ethiopian birr.
   */
  var ETB = {
    code: 'ETB',
    base: 10,
    exponent: 2
  };

  /**
   * Euro.
   */
  var EUR = {
    code: 'EUR',
    base: 10,
    exponent: 2
  };

  /**
   * Fiji dollar.
   */
  var FJD = {
    code: 'FJD',
    base: 10,
    exponent: 2
  };

  /**
   * Falkland Islands pound.
   */
  var FKP = {
    code: 'FKP',
    base: 10,
    exponent: 2
  };

  /**
   * Pound sterling.
   */
  var GBP = {
    code: 'GBP',
    base: 10,
    exponent: 2
  };

  /**
   * Georgian lari.
   */
  var GEL = {
    code: 'GEL',
    base: 10,
    exponent: 2
  };

  /**
   * Ghanaian cedi.
   */
  var GHS = {
    code: 'GHS',
    base: 10,
    exponent: 2
  };

  /**
   * Gibraltar pound.
   */
  var GIP = {
    code: 'GIP',
    base: 10,
    exponent: 2
  };

  /**
   * Gambian dalasi.
   */
  var GMD = {
    code: 'GMD',
    base: 10,
    exponent: 2
  };

  /**
   * Guinean franc.
   */
  var GNF = {
    code: 'GNF',
    base: 10,
    exponent: 0
  };

  /**
   * Guatemalan quetzal.
   */
  var GTQ = {
    code: 'GTQ',
    base: 10,
    exponent: 2
  };

  /**
   * Guyanese dollar.
   */
  var GYD = {
    code: 'GYD',
    base: 10,
    exponent: 2
  };

  /**
   * Hong Kong dollar.
   */
  var HKD = {
    code: 'HKD',
    base: 10,
    exponent: 2
  };

  /**
   * Honduran lempira.
   */
  var HNL = {
    code: 'HNL',
    base: 10,
    exponent: 2
  };

  /**
   * Croatian kuna.
   */
  var HRK = {
    code: 'HRK',
    base: 10,
    exponent: 2
  };

  /**
   * Haitian gourde.
   */
  var HTG = {
    code: 'HTG',
    base: 10,
    exponent: 2
  };

  /**
   * Hungarian forint.
   */
  var HUF = {
    code: 'HUF',
    base: 10,
    exponent: 2
  };

  /**
   * Indonesian rupiah.
   */
  var IDR = {
    code: 'IDR',
    base: 10,
    exponent: 2
  };

  /**
   * Israeli new shekel.
   */
  var ILS = {
    code: 'ILS',
    base: 10,
    exponent: 2
  };

  /**
   * Indian rupee.
   */
  var INR = {
    code: 'INR',
    base: 10,
    exponent: 2
  };

  /**
   * Iraqi dinar.
   */
  var IQD = {
    code: 'IQD',
    base: 10,
    exponent: 3
  };

  /**
   * Iranian rial.
   */
  var IRR = {
    code: 'IRR',
    base: 10,
    exponent: 2
  };

  /**
   * Icelandic króna.
   */
  var ISK = {
    code: 'ISK',
    base: 10,
    exponent: 0
  };

  /**
   * Jamaican dollar.
   */
  var JMD = {
    code: 'JMD',
    base: 10,
    exponent: 2
  };

  /**
   * Jordanian dinar.
   */
  var JOD = {
    code: 'JOD',
    base: 10,
    exponent: 3
  };

  /**
   * Japanese yen.
   */
  var JPY = {
    code: 'JPY',
    base: 10,
    exponent: 0
  };

  /**
   * Kenyan shilling.
   */
  var KES = {
    code: 'KES',
    base: 10,
    exponent: 2
  };

  /**
   * Kyrgyzstani som.
   */
  var KGS = {
    code: 'KGS',
    base: 10,
    exponent: 2
  };

  /**
   * Cambodian riel.
   */
  var KHR = {
    code: 'KHR',
    base: 10,
    exponent: 2
  };

  /**
   * Comoro franc.
   */
  var KMF = {
    code: 'KMF',
    base: 10,
    exponent: 0
  };

  /**
   * North Korean won.
   */
  var KPW = {
    code: 'KPW',
    base: 10,
    exponent: 2
  };

  /**
   * South Korean won.
   */
  var KRW = {
    code: 'KRW',
    base: 10,
    exponent: 0
  };

  /**
   * Kuwaiti dinar.
   */
  var KWD = {
    code: 'KWD',
    base: 10,
    exponent: 3
  };

  /**
   * Cayman Islands dollar.
   */
  var KYD = {
    code: 'KYD',
    base: 10,
    exponent: 2
  };

  /**
   * Kazakhstani tenge.
   */
  var KZT = {
    code: 'KZT',
    base: 10,
    exponent: 2
  };

  /**
   * Lao kip.
   */
  var LAK = {
    code: 'LAK',
    base: 10,
    exponent: 2
  };

  /**
   * Lebanese pound.
   */
  var LBP = {
    code: 'LBP',
    base: 10,
    exponent: 2
  };

  /**
   * Sri Lankan rupee.
   */
  var LKR = {
    code: 'LKR',
    base: 10,
    exponent: 2
  };

  /**
   * Liberian dollar.
   */
  var LRD = {
    code: 'LRD',
    base: 10,
    exponent: 2
  };

  /**
   * Lesotho loti.
   */
  var LSL = {
    code: 'LSL',
    base: 10,
    exponent: 2
  };

  /**
   * Libyan dinar.
   */
  var LYD = {
    code: 'LYD',
    base: 10,
    exponent: 3
  };

  /**
   * Moroccan dirham.
   */
  var MAD = {
    code: 'MAD',
    base: 10,
    exponent: 2
  };

  /**
   * Moldovan leu.
   */
  var MDL = {
    code: 'MDL',
    base: 10,
    exponent: 2
  };

  /**
   * Malagasy ariary.
   */
  var MGA = {
    code: 'MGA',
    base: 5,
    exponent: 1
  };

  /**
   * Macedonian denar.
   */
  var MKD = {
    code: 'MKD',
    base: 10,
    exponent: 2
  };

  /**
   * Myanmar kyat.
   */
  var MMK = {
    code: 'MMK',
    base: 10,
    exponent: 2
  };

  /**
   * Mongolian tögrög.
   */
  var MNT = {
    code: 'MNT',
    base: 10,
    exponent: 2
  };

  /**
   * Macanese pataca.
   */
  var MOP = {
    code: 'MOP',
    base: 10,
    exponent: 2
  };

  /**
   * Mauritanian ouguiya.
   */
  var MRU = {
    code: 'MRU',
    base: 5,
    exponent: 1
  };

  /**
   * Mauritian rupee.
   */
  var MUR = {
    code: 'MUR',
    base: 10,
    exponent: 2
  };

  /**
   * Maldivian rufiyaa.
   */
  var MVR = {
    code: 'MVR',
    base: 10,
    exponent: 2
  };

  /**
   * Malawian kwacha.
   */
  var MWK = {
    code: 'MWK',
    base: 10,
    exponent: 2
  };

  /**
   * Mexican peso.
   */
  var MXN = {
    code: 'MXN',
    base: 10,
    exponent: 2
  };

  /**
   * Mexican Unidad de Inversion.
   */
  var MXV = {
    code: 'MXV',
    base: 10,
    exponent: 2
  };

  /**
   * Malaysian ringgit.
   */
  var MYR = {
    code: 'MYR',
    base: 10,
    exponent: 2
  };

  /**
   * Mozambican metical.
   */
  var MZN = {
    code: 'MZN',
    base: 10,
    exponent: 2
  };

  /**
   * Namibian dollar.
   */
  var NAD = {
    code: 'NAD',
    base: 10,
    exponent: 2
  };

  /**
   * Nigerian naira.
   */
  var NGN = {
    code: 'NGN',
    base: 10,
    exponent: 2
  };

  /**
   * Nicaraguan córdoba.
   */
  var NIO = {
    code: 'NIO',
    base: 10,
    exponent: 2
  };

  /**
   * Norwegian krone.
   */
  var NOK = {
    code: 'NOK',
    base: 10,
    exponent: 2
  };

  /**
   * Nepalese rupee.
   */
  var NPR = {
    code: 'NPR',
    base: 10,
    exponent: 2
  };

  /**
   * New Zealand dollar.
   */
  var NZD = {
    code: 'NZD',
    base: 10,
    exponent: 2
  };

  /**
   * Omani rial.
   */
  var OMR = {
    code: 'OMR',
    base: 10,
    exponent: 3
  };

  /**
   * Panamanian balboa.
   */
  var PAB = {
    code: 'PAB',
    base: 10,
    exponent: 2
  };

  /**
   * Peruvian sol.
   */
  var PEN = {
    code: 'PEN',
    base: 10,
    exponent: 2
  };

  /**
   * Papua New Guinean kina.
   */
  var PGK = {
    code: 'PGK',
    base: 10,
    exponent: 2
  };

  /**
   * Philippine peso.
   */
  var PHP = {
    code: 'PHP',
    base: 10,
    exponent: 2
  };

  /**
   * Pakistani rupee.
   */
  var PKR = {
    code: 'PKR',
    base: 10,
    exponent: 2
  };

  /**
   * Polish złoty.
   */
  var PLN = {
    code: 'PLN',
    base: 10,
    exponent: 2
  };

  /**
   * Paraguayan guaraní.
   */
  var PYG = {
    code: 'PYG',
    base: 10,
    exponent: 0
  };

  /**
   * Qatari riyal.
   */
  var QAR = {
    code: 'QAR',
    base: 10,
    exponent: 2
  };

  /**
   * Romanian leu.
   */
  var RON = {
    code: 'RON',
    base: 10,
    exponent: 2
  };

  /**
   * Serbian dinar.
   */
  var RSD = {
    code: 'RSD',
    base: 10,
    exponent: 2
  };

  /**
   * Russian ruble.
   */
  var RUB = {
    code: 'RUB',
    base: 10,
    exponent: 2
  };

  /**
   * Rwandan franc.
   */
  var RWF = {
    code: 'RWF',
    base: 10,
    exponent: 0
  };

  /**
   * Saudi riyal.
   */
  var SAR = {
    code: 'SAR',
    base: 10,
    exponent: 2
  };

  /**
   * Solomon Islands dollar.
   */
  var SBD = {
    code: 'SBD',
    base: 10,
    exponent: 2
  };

  /**
   * Seychelles rupee.
   */
  var SCR = {
    code: 'SCR',
    base: 10,
    exponent: 2
  };

  /**
   * Sudanese pound.
   */
  var SDG = {
    code: 'SDG',
    base: 10,
    exponent: 2
  };

  /**
   * Swedish krona.
   */
  var SEK = {
    code: 'SEK',
    base: 10,
    exponent: 2
  };

  /**
   * Singapore dollar.
   */
  var SGD = {
    code: 'SGD',
    base: 10,
    exponent: 2
  };

  /**
   * Saint Helena pound.
   */
  var SHP = {
    code: 'SHP',
    base: 10,
    exponent: 2
  };

  /**
   * Sierra Leonean leone.
   */
  var SLL = {
    code: 'SLL',
    base: 10,
    exponent: 2
  };

  /**
   * Somali shilling.
   */
  var SOS = {
    code: 'SOS',
    base: 10,
    exponent: 2
  };

  /**
   * Surinamese dollar.
   */
  var SRD = {
    code: 'SRD',
    base: 10,
    exponent: 2
  };

  /**
   * South Sudanese pound.
   */
  var SSP = {
    code: 'SSP',
    base: 10,
    exponent: 2
  };

  /**
   * São Tomé and Príncipe dobra.
   */
  var STN = {
    code: 'STN',
    base: 10,
    exponent: 2
  };

  /**
   * Salvadoran colón.
   */
  var SVC = {
    code: 'SVC',
    base: 10,
    exponent: 2
  };

  /**
   * Syrian pound.
   */
  var SYP = {
    code: 'SYP',
    base: 10,
    exponent: 2
  };

  /**
   * Swazi lilangeni.
   */
  var SZL = {
    code: 'SZL',
    base: 10,
    exponent: 2
  };

  /**
   * Thai baht.
   */
  var THB = {
    code: 'THB',
    base: 10,
    exponent: 2
  };

  /**
   * Tajikistani somoni.
   */
  var TJS = {
    code: 'TJS',
    base: 10,
    exponent: 2
  };

  /**
   * Turkmenistan manat.
   */
  var TMT = {
    code: 'TMT',
    base: 10,
    exponent: 2
  };

  /**
   * Tunisian dinar.
   */
  var TND = {
    code: 'TND',
    base: 10,
    exponent: 3
  };

  /**
   * Tongan paʻanga.
   */
  var TOP = {
    code: 'TOP',
    base: 10,
    exponent: 2
  };

  /**
   * Turkish lira.
   */
  var TRY = {
    code: 'TRY',
    base: 10,
    exponent: 2
  };

  /**
   * Trinidad and Tobago dollar.
   */
  var TTD = {
    code: 'TTD',
    base: 10,
    exponent: 2
  };

  /**
   * New Taiwan dollar.
   */
  var TWD = {
    code: 'TWD',
    base: 10,
    exponent: 2
  };

  /**
   * Tanzanian shilling.
   */
  var TZS = {
    code: 'TZS',
    base: 10,
    exponent: 2
  };

  /**
   * Ukrainian hryvnia.
   */
  var UAH = {
    code: 'UAH',
    base: 10,
    exponent: 2
  };

  /**
   * Ugandan shilling.
   */
  var UGX = {
    code: 'UGX',
    base: 10,
    exponent: 0
  };

  /**
   * United States dollar.
   */
  var USD = {
    code: 'USD',
    base: 10,
    exponent: 2
  };

  /**
   * United States dollar (next day).
   */
  var USN = {
    code: 'USN',
    base: 10,
    exponent: 2
  };

  /**
   * Uruguay Peso en Unidades Indexadas.
   */
  var UYI = {
    code: 'UYI',
    base: 10,
    exponent: 0
  };

  /**
   * Uruguayan peso.
   */
  var UYU = {
    code: 'UYU',
    base: 10,
    exponent: 2
  };

  /**
   * Unidad previsional.
   */
  var UYW = {
    code: 'UYW',
    base: 10,
    exponent: 4
  };

  /**
   * Uzbekistani soʻm.
   */
  var UZS = {
    code: 'UZS',
    base: 10,
    exponent: 2
  };

  /**
   * Venezuelan bolívar.
   */
  var VES = {
    code: 'VES',
    base: 10,
    exponent: 2
  };

  /**
   * Vietnamese đồng.
   */
  var VND = {
    code: 'VND',
    base: 10,
    exponent: 0
  };

  /**
   * Vanuatu vatu.
   */
  var VUV = {
    code: 'VUV',
    base: 10,
    exponent: 0
  };

  /**
   * Samoan tālā.
   */
  var WST = {
    code: 'WST',
    base: 10,
    exponent: 2
  };

  /**
   * Central African CFA franc.
   */
  var XAF = {
    code: 'XAF',
    base: 10,
    exponent: 0
  };

  /**
   * East Caribbean dollar.
   */
  var XCD = {
    code: 'XCD',
    base: 10,
    exponent: 2
  };

  /**
   * West African CFA franc.
   */
  var XOF = {
    code: 'XOF',
    base: 10,
    exponent: 0
  };

  /**
   * CFP franc.
   */
  var XPF = {
    code: 'XPF',
    base: 10,
    exponent: 0
  };

  /**
   * Yemeni rial.
   */
  var YER = {
    code: 'YER',
    base: 10,
    exponent: 2
  };

  /**
   * South African rand.
   */
  var ZAR = {
    code: 'ZAR',
    base: 10,
    exponent: 2
  };

  /**
   * Zambian kwacha.
   */
  var ZMW = {
    code: 'ZMW',
    base: 10,
    exponent: 2
  };

  /**
   * Zimbabwean dollar.
   */
  var ZWL = {
    code: 'ZWL',
    base: 10,
    exponent: 2
  };

  exports.AED = AED;
  exports.AFN = AFN;
  exports.ALL = ALL;
  exports.AMD = AMD;
  exports.ANG = ANG;
  exports.AOA = AOA;
  exports.ARS = ARS;
  exports.AUD = AUD;
  exports.AWG = AWG;
  exports.AZN = AZN;
  exports.BAM = BAM;
  exports.BBD = BBD;
  exports.BDT = BDT;
  exports.BGN = BGN;
  exports.BHD = BHD;
  exports.BIF = BIF;
  exports.BMD = BMD;
  exports.BND = BND;
  exports.BOB = BOB;
  exports.BOV = BOV;
  exports.BRL = BRL;
  exports.BSD = BSD;
  exports.BTN = BTN;
  exports.BWP = BWP;
  exports.BYN = BYN;
  exports.BZD = BZD;
  exports.CAD = CAD;
  exports.CDF = CDF;
  exports.CHE = CHE;
  exports.CHF = CHF;
  exports.CHW = CHW;
  exports.CLF = CLF;
  exports.CLP = CLP;
  exports.CNY = CNY;
  exports.COP = COP;
  exports.COU = COU;
  exports.CRC = CRC;
  exports.CUC = CUC;
  exports.CUP = CUP;
  exports.CVE = CVE;
  exports.CZK = CZK;
  exports.DJF = DJF;
  exports.DKK = DKK;
  exports.DOP = DOP;
  exports.DZD = DZD;
  exports.EGP = EGP;
  exports.ERN = ERN;
  exports.ETB = ETB;
  exports.EUR = EUR;
  exports.FJD = FJD;
  exports.FKP = FKP;
  exports.GBP = GBP;
  exports.GEL = GEL;
  exports.GHS = GHS;
  exports.GIP = GIP;
  exports.GMD = GMD;
  exports.GNF = GNF;
  exports.GTQ = GTQ;
  exports.GYD = GYD;
  exports.HKD = HKD;
  exports.HNL = HNL;
  exports.HRK = HRK;
  exports.HTG = HTG;
  exports.HUF = HUF;
  exports.IDR = IDR;
  exports.ILS = ILS;
  exports.INR = INR;
  exports.IQD = IQD;
  exports.IRR = IRR;
  exports.ISK = ISK;
  exports.JMD = JMD;
  exports.JOD = JOD;
  exports.JPY = JPY;
  exports.KES = KES;
  exports.KGS = KGS;
  exports.KHR = KHR;
  exports.KMF = KMF;
  exports.KPW = KPW;
  exports.KRW = KRW;
  exports.KWD = KWD;
  exports.KYD = KYD;
  exports.KZT = KZT;
  exports.LAK = LAK;
  exports.LBP = LBP;
  exports.LKR = LKR;
  exports.LRD = LRD;
  exports.LSL = LSL;
  exports.LYD = LYD;
  exports.MAD = MAD;
  exports.MDL = MDL;
  exports.MGA = MGA;
  exports.MKD = MKD;
  exports.MMK = MMK;
  exports.MNT = MNT;
  exports.MOP = MOP;
  exports.MRU = MRU;
  exports.MUR = MUR;
  exports.MVR = MVR;
  exports.MWK = MWK;
  exports.MXN = MXN;
  exports.MXV = MXV;
  exports.MYR = MYR;
  exports.MZN = MZN;
  exports.NAD = NAD;
  exports.NGN = NGN;
  exports.NIO = NIO;
  exports.NOK = NOK;
  exports.NPR = NPR;
  exports.NZD = NZD;
  exports.OMR = OMR;
  exports.PAB = PAB;
  exports.PEN = PEN;
  exports.PGK = PGK;
  exports.PHP = PHP;
  exports.PKR = PKR;
  exports.PLN = PLN;
  exports.PYG = PYG;
  exports.QAR = QAR;
  exports.RON = RON;
  exports.RSD = RSD;
  exports.RUB = RUB;
  exports.RWF = RWF;
  exports.SAR = SAR;
  exports.SBD = SBD;
  exports.SCR = SCR;
  exports.SDG = SDG;
  exports.SEK = SEK;
  exports.SGD = SGD;
  exports.SHP = SHP;
  exports.SLL = SLL;
  exports.SOS = SOS;
  exports.SRD = SRD;
  exports.SSP = SSP;
  exports.STN = STN;
  exports.SVC = SVC;
  exports.SYP = SYP;
  exports.SZL = SZL;
  exports.THB = THB;
  exports.TJS = TJS;
  exports.TMT = TMT;
  exports.TND = TND;
  exports.TOP = TOP;
  exports.TRY = TRY;
  exports.TTD = TTD;
  exports.TWD = TWD;
  exports.TZS = TZS;
  exports.UAH = UAH;
  exports.UGX = UGX;
  exports.USD = USD;
  exports.USN = USN;
  exports.UYI = UYI;
  exports.UYU = UYU;
  exports.UYW = UYW;
  exports.UZS = UZS;
  exports.VES = VES;
  exports.VND = VND;
  exports.VUV = VUV;
  exports.WST = WST;
  exports.XAF = XAF;
  exports.XCD = XCD;
  exports.XOF = XOF;
  exports.XPF = XPF;
  exports.YER = YER;
  exports.ZAR = ZAR;
  exports.ZMW = ZMW;
  exports.ZWL = ZWL;

  Object.defineProperty(exports, '__esModule', { value: true });

}));
