// fast-check/lib/esm/fast-check-default.js
import { pre } from './check/precondition/Pre.js';
import { asyncProperty } from './check/property/AsyncProperty.js';
import { property } from './check/property/Property.js';
import { assert, check } from './check/runner/Runner.js';
import { sample, statistics } from './check/runner/Sampler.js';
import { gen } from './arbitrary/gen.js';
import { array } from './arbitrary/array.js';
import { bigInt } from './arbitrary/bigInt.js';
import { bigIntN } from './arbitrary/bigIntN.js';
import { bigUint } from './arbitrary/bigUint.js';
import { bigUintN } from './arbitrary/bigUintN.js';
import { boolean } from './arbitrary/boolean.js';
import { falsy } from './arbitrary/falsy.js';
import { ascii } from './arbitrary/ascii.js';
import { base64 } from './arbitrary/base64.js';
import { char } from './arbitrary/char.js';
import { char16bits } from './arbitrary/char16bits.js';
import { fullUnicode } from './arbitrary/fullUnicode.js';
import { hexa } from './arbitrary/hexa.js';
import { unicode } from './arbitrary/unicode.js';
import { constant } from './arbitrary/constant.js';
import { constantFrom } from './arbitrary/constantFrom.js';
import { context } from './arbitrary/context.js';
import { date } from './arbitrary/date.js';
import { clone } from './arbitrary/clone.js';
import { dictionary } from './arbitrary/dictionary.js';
import { emailAddress } from './arbitrary/emailAddress.js';
import { double } from './arbitrary/double.js';
import { float } from './arbitrary/float.js';
import { compareBooleanFunc } from './arbitrary/compareBooleanFunc.js';
import { compareFunc } from './arbitrary/compareFunc.js';
import { func } from './arbitrary/func.js';
import { domain } from './arbitrary/domain.js';
import { integer } from './arbitrary/integer.js';
import { maxSafeInteger } from './arbitrary/maxSafeInteger.js';
import { maxSafeNat } from './arbitrary/maxSafeNat.js';
import { nat } from './arbitrary/nat.js';
import { ipV4 } from './arbitrary/ipV4.js';
import { ipV4Extended } from './arbitrary/ipV4Extended.js';
import { ipV6 } from './arbitrary/ipV6.js';
import { letrec } from './arbitrary/letrec.js';
import { lorem } from './arbitrary/lorem.js';
import { mapToConstant } from './arbitrary/mapToConstant.js';
import { memo } from './arbitrary/memo.js';
import { mixedCase } from './arbitrary/mixedCase.js';
import { object } from './arbitrary/object.js';
import { json } from './arbitrary/json.js';
import { anything } from './arbitrary/anything.js';
import { unicodeJsonValue } from './arbitrary/unicodeJsonValue.js';
import { jsonValue } from './arbitrary/jsonValue.js';
import { unicodeJson } from './arbitrary/unicodeJson.js';
import { oneof } from './arbitrary/oneof.js';
import { option } from './arbitrary/option.js';
import { record } from './arbitrary/record.js';
import { uniqueArray } from './arbitrary/uniqueArray.js';
import { infiniteStream } from './arbitrary/infiniteStream.js';
import { asciiString } from './arbitrary/asciiString.js';
import { base64String } from './arbitrary/base64String.js';
import { fullUnicodeString } from './arbitrary/fullUnicodeString.js';
import { hexaString } from './arbitrary/hexaString.js';
import { string } from './arbitrary/string.js';
import { string16bits } from './arbitrary/string16bits.js';
import { stringOf } from './arbitrary/stringOf.js';
import { unicodeString } from './arbitrary/unicodeString.js';
import { subarray } from './arbitrary/subarray.js';
import { shuffledSubarray } from './arbitrary/shuffledSubarray.js';
import { tuple } from './arbitrary/tuple.js';
import { ulid } from './arbitrary/ulid.js';
import { uuid } from './arbitrary/uuid.js';
import { uuidV } from './arbitrary/uuidV.js';
import { webAuthority } from './arbitrary/webAuthority.js';
import { webFragments } from './arbitrary/webFragments.js';
import { webPath } from './arbitrary/webPath.js';
import { webQueryParameters } from './arbitrary/webQueryParameters.js';
import { webSegment } from './arbitrary/webSegment.js';
import { webUrl } from './arbitrary/webUrl.js';
import { commands } from './arbitrary/commands.js';
import { asyncModelRun, modelRun, scheduledModelRun } from './check/model/ModelRunner.js';
import { Random } from './random/generator/Random.js';
import { configureGlobal, readConfigureGlobal, resetConfigureGlobal, } from './check/runner/configuration/GlobalParameters.js';
import { VerbosityLevel } from './check/runner/configuration/VerbosityLevel.js';
import { ExecutionStatus } from './check/runner/reporter/ExecutionStatus.js';
import { cloneMethod, cloneIfNeeded, hasCloneMethod } from './check/symbols.js';
import { Stream, stream } from './stream/Stream.js';
import { hash } from './utils/hash.js';
import { stringify, asyncStringify, toStringMethod, hasToStringMethod, asyncToStringMethod, hasAsyncToStringMethod, } from './utils/stringify.js';
import { scheduler, schedulerFor } from './arbitrary/scheduler.js';
import { defaultReportMessage, asyncDefaultReportMessage } from './check/runner/utils/RunDetailsFormatter.js';
import { PreconditionFailure } from './check/precondition/PreconditionFailure.js';
import { int8Array } from './arbitrary/int8Array.js';
import { int16Array } from './arbitrary/int16Array.js';
import { int32Array } from './arbitrary/int32Array.js';
import { uint8Array } from './arbitrary/uint8Array.js';
import { uint8ClampedArray } from './arbitrary/uint8ClampedArray.js';
import { uint16Array } from './arbitrary/uint16Array.js';
import { uint32Array } from './arbitrary/uint32Array.js';
import { float32Array } from './arbitrary/float32Array.js';
import { float64Array } from './arbitrary/float64Array.js';
import { sparseArray } from './arbitrary/sparseArray.js';
import { Arbitrary } from './check/arbitrary/definition/Arbitrary.js';
import { Value } from './check/arbitrary/definition/Value.js';
import { createDepthIdentifier, getDepthContextFor } from './arbitrary/_internals/helpers/DepthContext.js';
import { bigInt64Array } from './arbitrary/bigInt64Array.js';
import { bigUint64Array } from './arbitrary/bigUint64Array.js';
import { stringMatching } from './arbitrary/stringMatching.js';
import { noShrink } from './arbitrary/noShrink.js';
import { noBias } from './arbitrary/noBias.js';
import { limitShrink } from './arbitrary/limitShrink.js';
const __type = 'module';
const __version = '3.23.2';
const __commitHash = 'a4a600eaa08c833707067a877db144289a724b91';
export { __type, __version, __commitHash, sample, statistics, check, assert, pre, PreconditionFailure, property, asyncProperty, boolean, falsy, float, double, integer, nat, maxSafeInteger, maxSafeNat, bigIntN, bigUintN, bigInt, bigUint, char, ascii, char16bits, unicode, fullUnicode, hexa, base64, mixedCase, string, asciiString, string16bits, stringOf, unicodeString, fullUnicodeString, hexaString, base64String, stringMatching, limitShrink, lorem, constant, constantFrom, mapToConstant, option, oneof, clone, noBias, noShrink, shuffledSubarray, subarray, array, sparseArray, infiniteStream, uniqueArray, tuple, record, dictionary, anything, object, json, jsonValue, unicodeJson, unicodeJsonValue, letrec, memo, compareBooleanFunc, compareFunc, func, context, gen, date, ipV4, ipV4Extended, ipV6, domain, webAuthority, webSegment, webFragments, webPath, webQueryParameters, webUrl, emailAddress, ulid, uuid, uuidV, int8Array, uint8Array, uint8ClampedArray, int16Array, uint16Array, int32Array, uint32Array, float32Array, float64Array, bigInt64Array, bigUint64Array, asyncModelRun, modelRun, scheduledModelRun, commands, scheduler, schedulerFor, Arbitrary, Value, cloneMethod, cloneIfNeeded, hasCloneMethod, toStringMethod, hasToStringMethod, asyncToStringMethod, hasAsyncToStringMethod, getDepthContextFor, stringify, asyncStringify, defaultReportMessage, asyncDefaultReportMessage, hash, VerbosityLevel, configureGlobal, readConfigureGlobal, resetConfigureGlobal, ExecutionStatus, Random, Stream, stream, createDepthIdentifier, };
